### PR TITLE
Intrepid2: vectorNorm fencing

### PR DESCRIPTION
**Pull request by MicheldeMessieres**
I'm not sure we'd want to do things this way, but opening the PR for discussion.

We need to fence the Serial vectorNorm for UVM. However the Serial vectorNorm is also used as the core of the kernel form, when we would not want to fence. We could resolve as in this PR, or fence every call of Serial vectorNorm which I don't think we'd want, or perhaps design this a different way.

Potentially this is a backwards compatibility issue if external code is using Serial vectorNorm in device kernels.

@trilinos/intrepid2 

## Motivation

To get Intrepid2 passing all unit tests with CUDA_LAUNCH_BLOCKING=0 in preparation for further work to analyze core fences in panzer.

## Testing

White Cuda Intrepid2 test suite.

### CODE REVIEW

**Comment by mperego** [comment]
@kyungjoo-kim do you mind having a look at this? Thanks!

File: packages/intrepid2/src/Shared/Intrepid2_RealSpaceTools.hpp [hunk 1]
```diff
@@ -93,7 +93,7 @@ namespace Intrepid2 {
       KOKKOS_INLINE_FUNCTION
       static inVecValueType
       vectorNorm( const Kokkos::DynRankView<inVecValueType,inVecProperties...> inVec, 
-                  const ENorm normType );
+                  const ENorm normType, bool bFence = true );
```
Comment (kyungjoo-kim) [inline packages/intrepid2/src/Shared/Intrepid2_RealSpaceTools.hpp:96]: This is Kokkos inline function which can be used inside of a functor. We cannot put a fence here.

File: packages/intrepid2/src/Shared/Intrepid2_RealSpaceToolsDef.hpp [hunk 2]
```diff
@@ -85,6 +85,11 @@ namespace Intrepid2 {
     const ordinal_type kend = inVec.extent(2);
     const ordinal_type lend = inVec.extent(3);
     const ordinal_type mend = inVec.extent(4);
+
+    if(bFence) {
+     Kokkos::fence();
+    }
```
Comment (kyungjoo-kim) [inline packages/intrepid2/src/Shared/Intrepid2_RealSpaceToolsDef.hpp:91]: fence here is not allowed.

**Comment by MicheldeMessieres** [comment]
@kyungjoo-kim There are a bunch of intrepid2 tests which use it directly as a host call.

For example Shared/ArrayTools/test_01.hpp has:

```
          if (rst::Serial::vectorNorm(out1_c_l_r, NORM_ONE) > tol) {
            *outStream << "\n\nINCORRECT contractFieldFieldScalar (1): check COMP_CPP vs. COMP_BLAS; "
                       << " diff-1norm = " << rst::Serial::vectorNorm(out1_c_l_r, NORM_ONE) << "\n\n";
            errorFlag = -1000;
          }
```
So the bFence was just to make sure the fence was only used when it was placed in the functor. I could also fix this by manually fencing all the unit tests where they call vectorNorm directly. Or maybe the unit tests should to be using it in this way.

**Comment by mperego** [comment]
Also here:
https://github.com/trilinos/Trilinos/blob/master/packages/intrepid2/src/Cell/Intrepid2_CellToolsDefPhysToRef.hpp#L180

https://github.com/trilinos/Trilinos/blob/master/packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceToolsDef.hpp#L510

**Comment by kyungjoo-kim** [comment]
Oh..... The test is wrong. The view should be mirrored into the host before it calculate the norm and check the value.

**Comment by mperego** [comment]
@kyungjoo-kim never mind my comment.. those are not using the serial version

**Comment by MicheldeMessieres** [comment]
@kyungjoo-kim Got it. Thanks. I will remove the internal fence stuff and fix the tests locally to mirror to host first.

**Comment by mperego** [comment]
@MicheldeMessieres @kyungjoo-kim OK, so we need to fix several tests. @MicheldeMessieres are you up for it, or would you like the @trilinos/intrepid2 to fix them?

**Comment by MicheldeMessieres** [comment]
@mperego I will take care of it.

**Comment by kyungjoo-kim** [comment]
Thanks. Those tests are a kind of direct translation of the old code a few years ago. At that time, our Kokkos-fu was so weak.

**Comment by MicheldeMessieres** [comment]
@mperego @kyungjoo-kim I've pushed an update which resolves all the Serial::vectorNorm calls to be on host. Also fixed a few device reads that were directly connected. There are probably more places throughout Trilinos tests that we are missing which are already protected by a fence somewhere. I expect we won't track all of those down until we run these tests on pure CudaSpace. But this will pass with CUDA_LAUNCH_BLOCKING=0.

Also I fixed some minor typos in output debug statements in test_02.hpp for ArrayTools. There were 2 spots printing data_c_p which should have been data_c_1.

**Comment by kyungjoo-kim** [comment]
@kddevin  @jwillenbring I have one question about testing Trilinos without cuda launch blocking. Are we going to have a PR test without cuda launch blocking so that we no longer introduce this type of errors after @MicheldeMessieres fixed all ?

**Comment by mperego** [comment]
@kyungjoo-kim yes, I think that's the plan.

**Comment by mperego** [comment]
> @mperego @kyungjoo-kim I've pushed an update which resolves all the Serial::vectorNorm calls to be on host. Also fixed a few device reads that were directly connected. There are probably more places throughout Trilinos tests that we are missing which are already protected by a fence somewhere. I expect we won't track all of those down until we run these tests on pure CudaSpace. But this will pass with CUDA_LAUNCH_BLOCKING=0.
> 
> Also I fixed some minor typos in output debug statements in test_02.hpp for ArrayTools. There were 2 spots printing data_c_p which should have been data_c_1.

@MicheldeMessieres Thanks Michel for fixing that!

**Pull request merged**